stats_reporter.go

```diff
@@ -69,26 +69,66 @@ type inMemoryCounter struct {
 	Tags  map[string]string
 }
 
-// getKey converts name+tags into a single string of the form
-// "name|tag1=value1|...|tagN=valueN", where tag names are sorted alphabetically.
-func (r *InMemoryStatsCollector) getKey(name string, tags map[string]string) string {
+// MetricDescr describes a metric with tags
+type MetricDescr struct {
+	Name string
+	Tags map[string]string
+}
+
+// NewMetricDescr creates a new metric descriptor from the name and an even number
+// of keyValues strings treates as alternating key, value pairs.
+func NewMetricDescr(name string, keyValues ...string) MetricDescr {
+	m := MetricDescr{Name: name}
+	if len(keyValues)%2 != 0 {
+		return m.WithTag("error", "uneven_keyValues_count")
+	}
+	for i := 0; i < len(keyValues); i += 2 {
+		m = m.WithTag(keyValues[i], keyValues[i+1])
+	}
+	return m
+}
+
+// WithTag returns a new metric descriptor with additional tag
+func (m MetricDescr) WithTag(key, value string) MetricDescr {
+	tags := m.Tags
+	if tags == nil {
```

Comment:
when you chain WithTag().WithTag(), the first time it's nil, the second time it's not